src/map.rs

```diff
@@ -0,0 +1,299 @@
+use crate::SysnoSet;
+use core::fmt;
+use std::fmt::Debug;
+use std::mem;
+
+/// A macro to simplify creation and optionally initializing a syscall map.
+///
+/// # Examples
+///
+/// Create an empty map with a default value of `0` and initialize it with some values:
+/// ```
+/// # use syscalls::{syscall_map, Sysno};
+/// let mut map = syscall_map!(
+///     0;
+///     Sysno::open => 0,
+///     Sysno::close => 42,
+/// );
+/// ```
+///
+/// Create a new map with the defaults set for a set of syscalls:
+/// ```
+/// # use syscalls::{syscall_map, Sysno, SysnoSet};
+/// let mut map = syscall_map!(0, SysnoSet::all());
+/// assert_eq!(map.count(), SysnoSet::all().count());
+/// ```
+///
+/// Use custom copy-iable type:
+/// ```
+/// # use syscalls::{syscall_map, Sysno};
+/// #[derive(Copy, Clone, Default)]
+/// struct Point { x: i32, y: i32 }
+/// let mut map = syscall_map!(Point::default());
+/// map.insert(Sysno::open, Point { x: 1, y: 2 });
+/// ```
+///
+/// Use function callbacks:
+/// ```
+/// # use syscalls::{syscall_map, Sysno, SysnoMap};
+/// type Handler = fn() -> i32;
+/// let map: SysnoMap<Option<Handler>> = syscall_map!(
+///     None::<Handler>;
+///     Sysno::open => Some(|| 1),
+///     Sysno::close => Some(|| -1),
+/// );
+///
+/// assert_eq!(map.get(Sysno::open).unwrap().unwrap()(), 1);
+/// assert_eq!(map.get(Sysno::close).unwrap().unwrap()(), -1);
+/// ```
+#[macro_export]
+macro_rules! syscall_map {
+    ($default:expr) => {
+        $crate::SysnoMap::new([$default; $crate::Sysno::table_size()])
+    };
+    ($default:expr, $syscall_set:expr) => {{
+        $crate::SysnoMap::new_with_set([$default; $crate::Sysno::table_size()], $syscall_set)
+    }};
+    ($default:expr; $($sysno:expr),* $(,)?) => {{
+        let mut map = $crate::syscall_map!($default);
+        $(
+            map.insert($sysno, $default);
+        )*
+        map
+    }};
+    ($default:expr; $($sysno:expr => $value:expr),* $(,)?) => {{
+        let mut map = $crate::syscall_map!($default);
+        $(
+            map.insert($sysno, $value);
+        )*
+        map
+    }};
+}
+
+/// A map of syscalls.
+///
+/// This provides constant-time lookup of syscalls within a bitset.
+///
+/// # Examples
+///
+/// ```
+/// # use syscalls::{syscall_map, Sysno, SysnoMap};
+/// let mut syscalls = syscall_map!(
+///     0;
+///     Sysno::open => 0,
+///     Sysno::close => 42,
+/// );
+/// assert!(!syscalls.is_empty());
+/// assert_eq!(syscalls.remove(Sysno::open), Some(0));
+/// assert_eq!(syscalls.insert(Sysno::close, 4), Some(42));
+/// assert!(syscalls.contains_key(Sysno::close));
+/// assert_eq!(syscalls.get(Sysno::close), Some(&4));
+/// assert_eq!(syscalls.insert(Sysno::close, 11), Some(4));
+/// assert_eq!(syscalls.count(), 1);
+/// assert_eq!(syscalls.remove(Sysno::close), Some(11));
+/// assert!(syscalls.is_empty());
+/// ```
+#[derive(Clone, Eq, PartialEq)]
+pub struct SysnoMap<T> {
+    is_set: SysnoSet,
+    data: [T; Sysno::table_size()],
```

Comment:
Is it possible to use `MaybeUninit<T>` here? It will require some `unsafe`, but I'm fine with that. It has a huge advantage of not requiring `T: Default`.

Reply:
I have never used `MaybeUninit<T>` so not certain of its limitations. If a user wants to use their own `Sized` type, and implement their own `Default` that is different from a simple zeroed-out memory, how would that work?

Reply:
So, `Default` really shouldn't be a requirement. The user can use it of course, if they wish.

With `MaybeUninit`, you just need to be careful to not call `.assume_init()` unless you're sure it is initialized. We can be sure that it is initialized if `is_set` contains the corresponding syscall.